### PR TITLE
feat: Add energy results to Unity Catalog

### PR DESCRIPTION
**Pull request by BjarkeMeier**
# Description

## Pull-request quality

- [x] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [ ] Tests are written and executed locally
- [x] Subsystem tests have been tested (by manually deploying to `dev_002`)

### CODE REVIEW

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py [hunk 1]
```diff
@@ -13,28 +13,175 @@
...
+        "energy_per_brp",
+        energy_per_brp,
+        WholesaleResultsInternalDatabase.ENERGY_PER_BRP_TABLE_NAME,
+        schemas.energy_per_brp_schema_uc,
+    )
+
+    # Write energy per energy supplier
+    energy_per_es = _union(
+        energy_results.production_per_ga_and_es,
+        energy_results.flex_consumption_per_ga_and_es,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:85]: not needed, right?

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py [hunk 2]
```diff
@@ -13,28 +13,175 @@
...
+        energy_results.flex_consumption_per_ga_and_brp_and_es,
+        energy_results.production_per_ga_and_brp_and_es,
+        energy_results.non_profiled_consumption_per_ga_and_es,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:88]: not needed, right?

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py (hunk 1 above)
Comment (JesperGomes) [reply source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:85]: same for production

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py [hunk 3]
```diff
@@ -13,28 +13,175 @@
...
+    grid_loss_metering_point_time_series = _union(
+        energy_results.positive_grid_loss,
+        energy_results.negative_grid_loss,
+    )
+    if grid_loss_metering_point_time_series:
+        grid_loss_metering_point_time_series = (
+            grid_loss_metering_point_time_series.withColumn(
+                EnergyResultColumnNames.metering_point_type,
+                f.when(
+                    f.col(EnergyResultColumnNames.quantity) > 0,
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:108]: or add the column directly on the dataframes before union?

Comment (JesperGomes) [reply source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:108]: isn't it also wrong? The consumption metering point will be assigned as production if quantity is zero

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py [hunk 4]
```diff
@@ -13,28 +13,175 @@
...
+) -> None:
     with logging_configuration.start_span(name):
 
         # Not all energy results have a value - it depends on the type of calculation
         if df is None:
             return None
+
+        # Adjust to match the schema
+        df = (
+            df.withColumn(
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/calculation/output/results/energy_results.py:159]: or withColumnRenamed?

File: source/databricks/calculation_engine/package/constants/energy_result_column_names.py [hunk 5]
```diff
@@ -18,13 +18,19 @@
 
 class EnergyResultColumnNames(ResultColumnNames):
     time_series_type = "time_series_type"
+    """Obsolete"""
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/constants/energy_result_column_names.py:21]: we decided it's not obsolete

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051201_create_energy_per_brp.sql [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    quantity_qualities ARRAY<STRING> NOT NULL
+)
+USING DELTA
+TBLPROPERTIES (
+    delta.deletedFileRetentionDuration = 'interval 30 days',
+    delta.constraints.calculation_id_chk = "LENGTH ( calculation_id ) = 36",
+    delta.constraints.result_id_chk = "LENGTH ( result_id ) = 36",
+    delta.constraints.grid_area_code_chk = "LENGTH ( grid_area_code ) = 3",
+    delta.constraints.balance_responsible_party_id_chk = "LENGTH ( balance_responsible_party_id ) = 13 OR LENGTH ( balance_responsible_party_id ) = 16",
+    delta.constraints.time_series_type_chk = "time_series_type IN ( 'production' , 'non_profiled_consumption' , 'net_exchange_per_ga' , 'flex_consumption' , 'grid_loss' , 'total_consumption' , 'temp_flex_consumption' , 'temp_production' )",
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051201_create_energy_per_brp.sql:23]: too many time_series_types

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051202_create_energy_per_es.sql [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+)
+USING DELTA
+TBLPROPERTIES (
+    delta.deletedFileRetentionDuration = 'interval 30 days',
+    delta.constraints.calculation_id_chk = "LENGTH ( calculation_id ) = 36",
+    delta.constraints.result_id_chk = "LENGTH ( result_id ) = 36",
+    delta.constraints.grid_area_code_chk = "LENGTH ( grid_area_code ) = 3",
+    delta.constraints.energy_supplier_id_chk = "LENGTH ( energy_supplier_id ) = 13 OR LENGTH ( energy_supplier_id ) = 16",
+    delta.constraints.balance_responsible_party_id_chk = "LENGTH ( balance_responsible_party_id ) = 13 OR LENGTH ( balance_responsible_party_id ) = 16",
+    delta.constraints.time_series_type_chk = "time_series_type IN ( 'production' , 'non_profiled_consumption' , 'net_exchange_per_ga' , 'flex_consumption' , 'grid_loss' , 'total_consumption' , 'temp_flex_consumption' , 'temp_production' )",
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051202_create_energy_per_es.sql:25]: too many time_series_type

File: source/databricks/calculation_engine/package/calculation/output/results/energy_results.py (hunk 3 above)
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/calculation/output/results/energy_results.py]: Much better, thx 👍

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051210_create_grid_loss_metering_point_time_series.sql [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+TBLPROPERTIES (
+    delta.deletedFileRetentionDuration = 'interval 30 days',
+    delta.constraints.calculation_id_chk = "LENGTH ( calculation_id ) = 36",
+    delta.constraints.result_id_chk = "LENGTH ( result_id ) = 36",
+    delta.constraints.grid_area_code_chk = "LENGTH ( grid_area_code ) = 3",
+    delta.constraints.energy_supplier_id_chk = "energy_supplier_id IS NULL OR LENGTH ( energy_supplier_id ) = 13 OR LENGTH ( energy_supplier_id ) = 16",
+    delta.constraints.balance_responsible_party_id_chk = "balance_responsible_party_id IS NULL OR LENGTH ( balance_responsible_party_id ) = 13 OR LENGTH ( balance_responsible_party_id ) = 16",
+    delta.constraints.metering_point_type_chk = "metering_point_type IN ( 'production' , 'consumption')",
+    delta.constraints.metering_point_id_chk = "metering_point_id IS NULL OR LENGTH ( metering_point_id ) = 18",
+    delta.constraints.metering_point_id_conditional_chk = "( time_series_type IN ( 'negative_grid_loss' , 'positive_grid_loss' ) AND metering_point_id IS NOT NULL ) OR ( time_series_type NOT IN ( 'negative_grid_loss' , 'positive_grid_loss' ) AND metering_point_id IS NULL )",
```
Comment (JesperGomes) [inline source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051210_create_grid_loss_metering_point_time_series.sql:28]: why does this include time_series_type?

Comment (JesperGomes) [reply source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051210_create_grid_loss_metering_point_time_series.sql:28]: and why is called metering_point_id?

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051210_create_grid_loss_metering_point_time_series.sql:28]: Copy paste :-|
Good catch 👍

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051202_create_energy_per_es.sql (hunk 7 above)
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051202_create_energy_per_es.sql]: Nice catch!

File: source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051201_create_energy_per_brp.sql (hunk 6 above)
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/package/datamigration/migration_scripts/202407051201_create_energy_per_brp.sql]: Nice catch!

**Review by JesperGomes — APPROVED** [review]

**Pull request merged**